relayer/workers/ethrelayer/ethereum-listener.go

```diff
@@ -47,63 +44,81 @@ func NewEthereumListener(
 		basicOutboundChannel:        nil,
 		incentivizedOutboundChannel: nil,
 		mapping:                     make(map[common.Address]string),
-		messages:                    messages,
-		headers:                     headers,
+		payloads:                    payloads,
+		headerSyncer:                nil,
 		log:                         log,
 	}
 }
 
 func (li *EthereumListener) Start(cxt context.Context, eg *errgroup.Group, initBlockHeight uint64, descendantsUntilFinal uint64) error {
+	closeWithError := func(err error) error {
+		li.log.Info("Shutting down listener...")
+		close(li.payloads)
+		return err
+	}
+
 	hcs, err := ethereum.NewHeaderCacheState(
 		eg,
 		initBlockHeight,
 		&ethereum.DefaultBlockLoader{Conn: li.conn},
 		nil,
 	)
 	if err != nil {
-		return err
+		return closeWithError(err)
 	}
 
 	basicOutboundChannel, err := outbound.NewBasicOutboundChannel(common.HexToAddress(li.config.Channels.Basic.Outbound), li.conn.GetClient())
 	if err != nil {
-		return err
+		return closeWithError(err)
 	}
 	li.basicOutboundChannel = basicOutboundChannel
 
 	incentivizedOutboundChannel, err := outbound.NewIncentivizedOutboundChannel(common.HexToAddress(li.config.Channels.Incentivized.Outbound), li.conn.GetClient())
 	if err != nil {
-		return err
+		return closeWithError(err)
 	}
 	li.incentivizedOutboundChannel = incentivizedOutboundChannel
 
 	li.mapping[common.HexToAddress(li.config.Channels.Basic.Outbound)] = "BasicInboundChannel.submit"
 	li.mapping[common.HexToAddress(li.config.Channels.Incentivized.Outbound)] = "IncentivizedInboundChannel.submit"
 
+	headersIn := make(chan *gethTypes.Header, 5)
+	li.headerSyncer = syncer.NewSyncer(
+		descendantsUntilFinal,
+		syncer.NewHeaderLoader(li.conn.GetClient()),
+		headersIn,
+		li.log,
+	)
+
 	eg.Go(func() error {
-		err := li.pollEventsAndHeaders(cxt, initBlockHeight, descendantsUntilFinal, hcs)
-		if li.messages != nil {
-			close(li.messages)
+		err := li.processEventsAndHeaders(cxt, initBlockHeight, descendantsUntilFinal, headersIn, hcs)
+
+		// Ensures the context is canceled so that the channel below is
+		// closed by the syncer
+		eg.Go(func() error { return err })
```

Comment:
Yes, it will. The problem is this: we need to wait for `headersIn` to close below before we can exit safely. But `headersIn` won't be closed unless the context is canceled.